src/components/select-field/Overlay.js

```diff
@@ -0,0 +1,127 @@
+// @flow
+import * as React from 'react';
+import classNames from 'classnames';
+import { getDimensions, getViewportIntersections } from '../../utils/dom';
+import { OVERLAY_WRAPPER_CLASS, VIEWPORT_BORDERS } from '../../constants';
+
+type Props = {
+    children: React.Node,
+    className?: string,
+    isOpen?: boolean,
+};
+
+type Dimensions = {
+    height?: number,
+    width?: number,
+};
+
+type Translations = {
+    x?: number,
+    y?: number,
+};
+
+function getDimensionsWhileHidden(element: HTMLElement): Dimensions {
+    if (!element) {
+        return {};
+    }
+
+    element.classList.add('offscreen');
+    const dimensions = getDimensions(element);
+    element.classList.remove('offscreen');
+    return dimensions;
+}
+
+class Overlay extends React.Component<Props, State> {
+    overlayRef: ?HTMLDivElement;
+
+    buttonRef: ?HTMLButtonElement;
+
+    constructor(props) {
+        super(props);
+        this.buttonRef = React.createRef();
+        this.overlayRef = React.createRef();
+    }
+
+    getOverlayContentDimensions = (): Dimensions => {
+        const node = this.overlayRef.current;
+        const content = node && node.firstChild;
+
+        const dimensions = content.clientHeight ? getDimensions(content) : getDimensionsWhileHidden(content);
+
+        return dimensions;
+    };
+
+    getOverlayIntersections = (): Array<string> => {
+        const { isOpen } = this.props;
+        const node = this.overlayRef.current;
+
+        if (!isOpen) {
+            return false;
+        }
+
+        const dimensions = this.getOverlayContentDimensions();
+        const boundingClientRect = node && node.getBoundingClientRect();
+        const bounding = {
+            top: boundingClientRect.top,
+            left: boundingClientRect.left,
+            bottom: boundingClientRect.top + dimensions.height,
+            right: boundingClientRect.left + dimensions.width,
+        };
+
+        return getViewportIntersections(bounding);
+    };
+
+    getOverlayTranslations = (intersections: Array<string>): Translations => {
+        const { height: overlayHeight, width: overlayWidth } = this.getOverlayContentDimensions();
+        const { height: buttonHeight, width: buttonWidth } = this.buttonRef.current.getBoundingClientRect();
+        const translations = { x: 0, y: 0 };
+
+        if (intersections.length >= 3) {
```

Comment:
Can you add comment why returning from here? May be add a constant for value `3`? Not sure why we return if more than 2 sides of rectangle intersect with viewport.

Reply:
I'll add a comment, but basically my thinking is that if there are more than 2 intersections with the viewport, part of the dropdown will appear outside the viewport regardless, so there is no need to move the overlay around